### PR TITLE
PipelineDesignPattern Project

### PR DESCRIPTION
**Pull request by aminafra**

### CODE REVIEW

File: src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace PipelineDesignPattern;
+public class Framework
+{
+    public void Authentication(HttpContext httpContext, Action<HttpContext> action)
+    {
+		"Start Authentication...".Dump();
+		if (httpContext.Country is CountryIPPAddress.Iran)
+		{
+			throw new Exception(message: "Iranian IPs are blocked");
```
Comment (mohammadKarimi) [inline src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs:11]: Throwing a generic Exception is not recommended. It's better to create a custom exception type that provides more information about the error.

File: src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+﻿using Dumpify;
+
+namespace PipelineDesignPattern;
+public class Framework
+{
+    public void Authentication(HttpContext httpContext, Action<HttpContext> action)
+    {
+		"Start Authentication...".Dump();
+		if (httpContext.Country is CountryIPPAddress.Iran)
```
Comment (mohammadKarimi) [inline src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs:9]: There is no null check for the httpContext parameter in both methods.

Comment (mohammadKarimi) [inline src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs:9]: if (httpContext is null)
        {
                throw new ArgumentNullException(nameof(httpContext));
        }

File: src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs (hunk 1 above)
Comment (mohammadKarimi) [inline src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs:11]: public class IranianIPBlockedException : Exception
{
    public IranianIPBlockedException(string message) : base(message)
    {
    }
}

File: src/Amin Afrasiabi/PipelineDesignPattern/HttpContext.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿namespace PipelineDesignPattern;
+public class HttpContext
+{
+    public CountryIPPAddress Country => Enum.TryParse(IP[^2..], out CountryIPPAddress country) ? country : 0;
+    public required string IP { get; set; }
+}
+public enum CountryIPPAddress
```
Comment (mohammadKarimi) [inline src/Amin Afrasiabi/PipelineDesignPattern/HttpContext.cs:7]: Naming convention, 
IPP ??

File: src/Amin Afrasiabi/PipelineDesignPattern/HttpContext.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿namespace PipelineDesignPattern;
+public class HttpContext
+{
+    public CountryIPPAddress Country => Enum.TryParse(IP[^2..], out CountryIPPAddress country) ? country : 0;
+    public required string IP { get; set; }
```
Comment (mohammadKarimi) [inline src/Amin Afrasiabi/PipelineDesignPattern/HttpContext.cs:5]: required 👍

**Review by mohammadKarimi — CHANGES_REQUESTED** [review]

**Comment by aminafra** [comment]
I tried to fix these issues

File: src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+﻿using Dumpify;
+
+namespace PipelineDesignPattern;
+public class Framework
+{
+    public void Authentication(HttpContext httpContext, Action<HttpContext> action)
+    {
+        ArgumentNullException.ThrowIfNull(httpContext);
```
Comment (mohammadKarimi) [inline src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs:8]: Perfect.

File: src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+        "Start Authentication...".Dump();
+        if (httpContext.Country is CountryIPAddress.Iran)
```
Comment (mohammadKarimi) [inline src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs:11]: constant pattern matching 👍

File: src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            throw new IranianIPBlockedException(message: "Iranian IPs are blocked");
```
Comment (mohammadKarimi) [inline src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs:13]: Custom Exception 👍

File: src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+    public void ExceptionHandling(HttpContext httpContext, Action<HttpContext> action)
+    {
+        try
+        {
+            ArgumentNullException.ThrowIfNull(httpContext);
+            action(httpContext);
+            "Done".Dump();
+        }
+        catch (Exception ex)
```
Comment (mohammadKarimi) [inline src/Amin Afrasiabi/PipelineDesignPattern/Framework.cs:26]: this block catches all type of Exception. it is better approach to catch issue related.

**Review by mohammadKarimi — APPROVED** [review]
I will merge your code. but consider you shouldn't catch all Exceptions, in this scenario.

**Pull request merged**